internal/app/machined/pkg/controllers/runtime/extension_services_config_files.go

```diff
@@ -0,0 +1,153 @@
+type ExtensionServicesConfigFilesController struct {
+	V1Alpha1Mode            v1alpha1runtime.Mode
+	V1Alpha1Services        ServiceManager
+	ExtensionsConfigBaseDir string
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *ExtensionServicesConfigFilesController) Name() string {
+	return "runtime.ExtensionServicesConfigFilesController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *ExtensionServicesConfigFilesController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.ExtensionServicesConfigType,
+			Kind:      controller.InputStrong,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *ExtensionServicesConfigFilesController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *ExtensionServicesConfigFilesController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	if ctrl.V1Alpha1Mode == v1alpha1runtime.ModeContainer {
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		list, err := safe.ReaderListAll[*runtime.ExtensionServicesConfig](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing extension services config: %w", err)
+		}
+
+		touchedIDs := make(map[string]struct{}, list.Len())
+
+		for iter := list.Iterator(); iter.Next(); {
+			extensionConfigPath := filepath.Join(ctrl.ExtensionsConfigBaseDir, iter.Value().Metadata().ID())
+
+			if err := os.MkdirAll(extensionConfigPath, 0o755); err != nil {
+				return fmt.Errorf("error creating directory %q: %w", extensionConfigPath, err)
+			}
+
+			touchedIDs[extensionConfigPath] = struct{}{}
+
+			for _, file := range iter.Value().TypedSpec().Files {
+				fileName := filepath.Join(extensionConfigPath, strings.ReplaceAll(strings.TrimPrefix(file.MountPath, "/"), "/", "-"))
+
+				content, err := os.ReadFile(fileName)
+				if err != nil {
+					if !os.IsNotExist(err) {
+						return fmt.Errorf("error reading file %q: %w", fileName, err)
+					}
+
+					if err = os.WriteFile(fileName, []byte(file.Content), 0o644); err != nil {
+						return fmt.Errorf("error writing file %q: %w", fileName, err)
+					}
+				} else {
+					if val := bytes.Compare(content, []byte(file.Content)); val != 0 {
+						if err = os.WriteFile(fileName, []byte(file.Content), 0o644); err != nil {
+							return fmt.Errorf("error writing file %q: %w", fileName, err)
+						}
+
+						if err = ctrl.handleRestart(ctx, "ext-"+iter.Value().Metadata().ID()); err != nil {
+							return fmt.Errorf("error restarting extension service: %w", err)
+						}
+					}
+				}
+
+				touchedIDs[fileName] = struct{}{}
+			}
+		}
+
+		// cleanup
+		if err := filepath.WalkDir(ctrl.ExtensionsConfigBaseDir, func(path string, d fs.DirEntry, err error) error {
```

Comment:
the ext service is not restarted on config file removal, that still needs to be fixed

Reply:
resolved